### PR TITLE
[Frontend] Logic to transform array of Datapoints into a table, display Review Metrics page with basic styling

### PR DESCRIPTION
**Pull request by terryttsai**
## Description of the change

Implement logic to transform an array of Datapoints into a table, display Review Metrics page with incomplete styling (to be finalized in a follow-up PR).

For now, I am hydrating the data from the Datapoints API. In a follow-up, we will get the Datapoints from the Data Upload API instead.

To see this page, visit http://localhost:3000/review_metrics

NOTE:
- The display order of the metrics, disaggregations, and dimensions are different than the metric config or data entry page. We are discussing ways of maintaining the same order: https://recidiviz.slack.com/archives/C02VDN1LSSW/p1663097034150579

Screen Shot 2022-09-13 at 3 43 06 PM

## Type of change

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #XXXX

### Development

**This box MUST be checked by the submitter prior to merging**:
- [ ] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  const renderSection = (
+    metricName: string,
+    datapoints: DatapointsGroupedByAggregateAndDisaggregations,
+    index: number
+  ) => {
+    return (
+      <SectionContainer key={metricName}>
+        <SectionTitleContainer>
+          <SectionTitleNumber>{index + 1}</SectionTitleNumber>
+          <SectionTitle>{metricName}</SectionTitle>
```
Comment (terryttsai) [inline publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:82]: My datapoints API only has the metric key, not the metric name, but Nichelle's Data Upload API will include the metric name.

**Comment by lilidworkin** [comment]
@terryttsai I think this branch will need to be rebased, and it might be a pain in the butt -- please let me know if it turns into a total disaster ... ! I will review it after that! Thanks for putting this up

**Comment by terryttsai** [comment]
@lilidworkin I used the command `git rebase -X ours main` to automatically resolve merge conflicts by siding with `main` and it looks everything resolved beautifully, though there's that extra line in the README for some reason again 😆

**Review by terryttsai — COMMENTED** [review]
A note on reviewing: Nichelle and I discussed the API and Nichelle is going to work to work on a new API that returns not an array of datapoints, but datapoints keyed by metric and disaggregations, so a lot of this logic will go away https://paper.dropbox.com/doc/API-Bulk-Upload-46N7XyKeuFHmRvZEUczqg

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+     */
+
+    // create a mapping from start date to the column index the start date is located in for fast lookup
+    const startDatesIndexLookup = startDates.reduce((map, current, idx) => {
+      map[current] = idx; /* eslint-disable-line no-param-reassign */
+      return map;
+    }, {} as { [key: string]: number });
+
+    // Create array of aggregate values, each value indexed with their corresponding date column
+    const aggregateRowData: DatapointValue[] = [];
```
Comment (terryttsai) [inline publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:116]: Specifically, L116-146 to generate the mapping of metrics and disaggregations will go away, and possibly getStartDates and startDatesIndexLookup might go away too if the dates are in order from the server, which they likely will be.

Comment (terryttsai) [reply publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:116]: The logic to focus on is really just the render logic in renderSection, renderDatapointsTable, the main render return body, and also metricKeyToDisplayName

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  SectionTitleContainer,
+  SectionTitleMonths,
+  SectionTitleNumber,
+  Subheading,
+} from "./ReviewMetrics.styles";
+
+const ReviewMetrics: React.FC = observer(() => {
+  const { userStore, datapointsStore } = useStore();
+
+  useEffect(
```
Comment (terryttsai) [inline publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:60]: This useEffect to fetch datapoints will go away once I'm hooked up to the Data Upload API

Comment (lilidworkin) [reply publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:60]: Makes sense!

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+      </SectionContainer>
+    );
+  };
+
+  const renderDatapointsTable = (
+    datapoints: DatapointsGroupedByAggregateAndDisaggregations,
+    startDates: string[]
+  ) => {
+    /**
+     * The Datapoints Table is made up of a few parts:
```
Comment (lilidworkin) [inline publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:99]: Love this docstring!

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx (hunk 2 above)
Comment (lilidworkin) [reply publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:116]: Cool yep, makes sense! Thanks for calling this out

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+          Before publishing, take a moment to review the changes. If you believe
+          there is an error, please contact the Justice Counts team via{" "}
+          <a href="mailto:support@justice-counts.org">
+            support@justicecounts.org
+          </a>
+        </Subheading>
+        <Divider />
+        {Object.entries(datapointsStore.datapointsByMetric).map(
+          ([metricName, datapoints], idx) => {
+            return renderSection(metricName, datapoints, idx);
```
Comment (lilidworkin) [inline publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:225]: Love how this component is organized, all feels super clean

**Review by lilidworkin — COMMENTED** [review]
@terryttsai this looks great to me! Organization is really clear, and I appreciate all the inline comments. Don't think I have any suggestions at all!

If you'd like to merge it as-is, and then continue working on top of it, that's fine with me and I'll accept. Otherwise if you'd like to make the changes based on the new API first, I can re-review after that. Either way, I think it's fine to merge before we hook up to the new API (i.e. while you're still hydrating from the Datapoint API). That will let us focus on one thing at a time. LMK if this all makes sense!

File: publisher/src/App.tsx [hunk 6]
```diff
@@ -47,6 +48,7 @@ const App: React.FC = (): ReactElement => {
           <Route path="/settings" element={<AccountSettings />} />
           <Route path="/metrics" element={<MetricsView />} />
           <Route path="/upload" element={<DataUpload />} />
+          <Route path="/review_metrics" element={<ReviewMetrics />} />
```
Comment (mxosman) [inline publisher/src/App.tsx:51]: nit: I don't know if there's a standard with this one - but I feel like I mostly see kabab case in route paths vs snake case.

File: publisher/src/components/DataViz/utils.ts [hunk 7]
```diff
@@ -333,3 +334,25 @@ export const getLatestDateFormatted = (
   }
   return "N/A";
 };
+
+export const getStartDates = (
+  data: DatapointsGroupedByAggregateAndDisaggregations
+) => {
+  const startDateSet = new Set<string>();
```
Comment (mxosman) [inline publisher/src/components/DataViz/utils.ts:341]: Curious what inspired you to use a Set? I need to do a better job exploring these built-ins!

File: publisher/src/components/ReviewMetrics/ReviewMetrics.styles.tsx [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+  overflow: hidden;
+  text-overflow: ellipsis;
+  white-space: nowrap;
+`;
+
+export const DatapointsTableDetailsContainer = styled.div`
+  overflow-x: scroll;
+`;
+export const DatapointsTableDetailsTable = styled.table``;
+export const DatapointsTableDetailsRow = styled.tr``;
```
Comment (mxosman) [inline publisher/src/components/ReviewMetrics/ReviewMetrics.styles.tsx:143]: Awesome! Yeah - I never know what to put in here when I just need to abstract it into a styled component and nothing else.

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+        () => userStore.userInfoLoaded,
+        () => {
+          datapointsStore.getDatapoints();
+        }
+      ),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    []
+  );
+
+  const renderSection = (
```
Comment (mxosman) [inline publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:73]: What is Section referring to? A metric?

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx (hunk 4 above)
Comment (mxosman) [reply publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:99]: Super helpful!

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+            if (!disaggregationRowData[disaggregation][dimension]) {
+              disaggregationRowData[disaggregation][dimension] = [];
+            }
+            disaggregationRowData[disaggregation][dimension][
+              startDatesIndexLookup[datapoint.start_date]
+            ] = value;
+          });
+        });
+      }
+    );
```
Comment (mxosman) [inline publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:146]: Wow!

File: publisher/src/stores/DatapointsStore.ts [hunk 11]
```diff
@@ -48,6 +48,14 @@ class DatapointsStore {
     this.loading = true;
   }
 
+  get metricKeyToDisplayName(): { [metricKey: string]: string | null } {
+    const mapping: { [metricKey: string]: string | null } = {};
+    this.rawDatapoints.forEach((dp) => {
+      mapping[dp.metric_definition_key] = dp.metric_display_name;
+    });
+    return mapping;
+  }
```
Comment (mxosman) [inline publisher/src/stores/DatapointsStore.ts:57]: Nice!

**Comment by mxosman** [comment]
This is awesome, Terry! Thank you for all of your helpful comments.

Just a couple of things that popped up - you might already be taking care of these in the follow ups:
- I think you'll need to wrap a <tbody></tbody> around all of the <tr> tags. There's a nice explanation here for that and will get rid of the console warning: https://github.com/facebook/react/issues/5652
- All of the components that are being rendered from mapping should have a unique key (specifically in the `renderDatapointsTable`)

Also, super curious to see how you're thinking about hydrating from the Data Upload API.

File: publisher/src/App.tsx (hunk 6 above)
Comment (terryttsai) [reply publisher/src/App.tsx:51]: You're right, looks like kebab case is the gold standard! I'll update

File: publisher/src/components/ReviewMetrics/ReviewMetrics.tsx (hunk 9 above)
Comment (terryttsai) [reply publisher/src/components/ReviewMetrics/ReviewMetrics.tsx:73]: Yeah, I'll rename this to `renderMetricSection`

File: publisher/src/components/DataViz/utils.ts (hunk 7 above)
Comment (terryttsai) [reply publisher/src/components/DataViz/utils.ts:341]: I think it was just more of how I wanted to present the code logic: I could use an object or a Map, and set the key as the start date, but what would I set for the value? I could just set the value as "True", but it's not used, so a Set accomplishes the same thing without needing to set an arbitrary value.

**Comment by terryttsai** [comment]
For the actual values of each cell on the table, some reports won't contain a datapoint for every dimension, so in that case I render an empty <td/>, what do you think the key should be in that case? There's no datapoint associated with it to have some sort of unique key to refer to. 

I just made an eslint exception for that case, since there is no guaranteed uniqueness in these values.

**Comment by mxosman** [comment]
> For the actual values of each cell on the table, some reports won't contain a datapoint for every dimension, so in that case I render an empty , what do you think the key should be in that case? There's no datapoint associated with it to have some sort of unique key to refer to.
> 
> I just made an eslint exception for that case, since there is no guaranteed uniqueness in these values.

Great call! In those case where there's no unique key, some times I use the index? But I know that's not always wise because it can cause some issues.

File: publisher/src/components/DataViz/utils.ts (hunk 7 above)
Comment (mxosman) [reply publisher/src/components/DataViz/utils.ts:341]: Ah - thank you for running me through your thought process! Makes perfect sense.

**Review by mxosman — APPROVED** [review]
LGTM!

**Comment by terryttsai** [comment]
Oh yeah, my bad I mean by making an eslint exception, I am keying this case by array index, and adding an eslint ignore about using the array index

**Comment by mxosman** [comment]
> Oh yeah, my bad I mean by making an eslint exception, I am keying this case by array index, and adding an eslint ignore about using the array index

Oh gotcha! That makes sense! Great call!

**Comment by terryttsai** [comment]
Thanks all for the comments!

**Pull request merged**